### ISSUE TITLE
Use Net MQ for compatibility with Linux / Mono

### ISSUE THREAD

**Issue by prosconi** [BayardRock/IfSharp#41]
The current solution for zmq is to use fszmq which uses native binaries. Net MQ is a pure .NET implementation of zmq.

**Comment by pblasucci** [comment]
For what it's worth, I'm not sure what switching to NetMQ saves you. The native binaries required by fszmq are the same ones required by PyZmq. So they'll need to be present anyway.

**Comment by mndrake** [comment]
@pblasucci The issues that I've had in the past with zmq have been with the path resolution of the zmq  (dll/so/dylib) that mono uses which is different on Windows versus OS X versus Linux (http://www.mono-project.com/docs/advanced/pinvoke/). If using Anaconda I've never had issues with the PyZmq - zmq resolution, it's just the .Net/Mono - zmq resolution that can be frustrating across platforms. I definitely defer to you though on expertise around zmq. I like the API that has been built within fszmq, I just don't like how mono finds the native libraries across platforms.

**Comment by pblasucci** [comment]
@mndrake I understand the grief cross-platform path resolution can cause. Fortunately, the information on the page you listed appears to be out-of-date. Recent versions of Mono (at least on OS X) exhibit the same behavior as .NET does on Windows. Specifically, the easiest set up is: 

+ put the native `.dll` in the local `bin`

or, failing that

+ put libzmq in a location which is part of the `PATH` environment variable, 

additionally, ensure

+ you've included `fszmq.dll.config` (with the correct info) alongside `fszmq.dll`

While I've not tested this on Linux, I'm certain this works on OS X (which is, in fact, the primary OS for fszmq development). Additionally, the latest version of fszmq (`4.0.9-prerelease`) eliminates the need for a specially-compiled build for BSD-like networking stacks. 

_(Also, as an aside, the easiest way to install libzmq on OS X is to use Homebrew.)_

**Comment by mndrake** [comment]
Path resolution in OS X is actually where I had the problem last year. Windows and Linux would search the local bin folder first, but OS X would ignore. If this is no longer the case it sounds like the required libzmq.(so/dylib/dll) could just be included in the bin folder (like it is currently for Windows/.NET)? ZMQ is only being used for 3-4 methods and client only, are there advantages for using the native libraries for the client/kernel side?

For sections of the project that require platform specific logic we could add custom compiler directives which I've seen used in code for mono specific logic.

```fsharp
#if MONO
// do something
#else
// do something else
#endif
```

Which I believe can also be used within project files, so could be used to swap between platform specific references like `libzmq` or`FSharp.Charting` depending on target used for building.

```
<Reference Include="Some.Assembly" Condition="" '$(Framework)' == 'NET20' " >
  <HintPath>..\Lib\$(Framework)\Some.Assembly.dll</HintPath>
</Reference>
```

Would this be a bad design choice and just over complicate things?

http://stackoverflow.com/questions/2923210/conditional-compilation-and-framework-targets
https://msdn.microsoft.com/en-us/library/dd233195.aspx

_( Also, as an aside, Homebrew is awesome unless you try to installed over a botched old version of Fink and Macports ;) )_

**Comment by pblasucci** [comment]
Yeah, the change on OS X (to look in the local bin) happened near the end of 2014 (sometime around version 3.10.0 of Mono. And it wasn't publicized... I found it totally by accident.   :-)

As for conditional compilation, I try to avoid it as much as possible (outside of `.fsx` files). However, it's sometimes necessary. 

With respect to project files, I've found MSBuild (or more likely Visual Studio and Xamarin Studio) to be fairly flakey. So, the approach I've started using in `fszmq` is to have two separate `.fsproj` files (with one actual `.fs` file shared between them). This makes it easy to associate all the Windows version OS X stuff.
